resolwe_bio/processes/variant_calling/wgs_preprocess_bwa2.py

```diff
@@ -316,56 +424,103 @@ def run(self, inputs, outputs):
         # File cleanup
         Path(sorted_bam).unlink(missing_ok=True)
 
-        # BaseRecalibrator
-        base_recal_inputs = [
+        # Prepare files for scattering over chromosomes
+        intervals_path = Path("intervals_folder")
+        intervals_path.mkdir(exist_ok=True)
+
+        if inputs.advanced_options.n_jobs:
+            n_jobs = max(inputs.advanced_options.n_jobs, 1)
+        else:
+            n_jobs = max(self.requirements.resources.cores, 1)
+
+        chromosome_sizes = "chromosome_sizes.bed"
+        prepare_chromosome_sizes(
+            fai_path=inputs.ref_seq.output.fai.path, bed_path=chromosome_sizes
+        )
+
+        split_intervals_inputs = [
             "-R",
             inputs.ref_seq.output.fasta.path,
-            "-I",
-            sorted_rg,
-            "--use-original-qualities",
-            "--tmp-dir",
-            TMPDIR,
+            "-L",
+            chromosome_sizes,
+            "--scatter-count",
+            n_jobs,
+            "--subdivision-mode",
+            "BALANCING_WITHOUT_INTERVAL_SUBDIVISION_WITH_OVERFLOW",
             "-O",
-            recal_table,
+            str(intervals_path),
         ]
-        # Add known sites to the input parameters of BaseRecalibrator.
-        for site in inputs.known_sites:
-            base_recal_inputs.extend(["--known-sites", f"{site.output.vcf.path}"])
-
-        return_code, _, _ = Cmd["gatk"]["BaseRecalibrator"][base_recal_inputs] & TEE(
+        return_code, _, _ = Cmd["gatk"]["SplitIntervals"][split_intervals_inputs] & TEE(
             retcode=None
         )
         if return_code:
-            self.error("BaseRecalibrator analysis failed.")
+            self.error("SplitIntervals tool failed.")
+
+        # BaseRecalibrator
+        recal_dir = Path("recal_tables")
+        recal_dir.mkdir(exist_ok=True)
+
+        intervals = [path for path in intervals_path.glob("*.interval_list")]
+        return_codes = Parallel(n_jobs=n_jobs)(
+            run_base_recalibration(
+                intput_bam=sorted_rg,
+                known_sites=inputs.known_sites,
+                interval_path=interval_path,
+                ref_seq_path=inputs.ref_seq.output.fasta.path,
+                tmp=TMPDIR,
+                parent_dir=recal_dir,
+            )
+            for interval_path in intervals
+        )
+        if any(return_codes):
+            self.error("GATK BaseRecalibrator tool failed.")
+
+        gather_bqsr_inputs = [
+            "-O",
+            recal_table,
+            *prepare_scattered_inputs(results_dir=recal_dir, pattern="*.csv"),
```

Comment:
Does this work? `*prepare_scattered_inputs(...)`

Reply:
It should work, but it can also be changed to `expand(...)` which is more common in our codebase.

```
In [36]: gather_bqsr_inputs = [ 
    ...:             "-O", 
    ...:             "recal_table", 
    ...:             *prepare_scattered_inputs(results_dir=Path("intervals_folder"), pattern="*.interval_list"), 
    ...:         ]                                                                                                                                                                                                                                                                                         

In [37]: gather_bqsr_inputs                                                                                                                                                                                                                                                                                
Out[37]: 
['-O',
 'recal_table',
 '-I',
 PosixPath('intervals_folder/0000-scattered.interval_list'),
 '-I',
 PosixPath('intervals_folder/0001-scattered.interval_list'),
 '-I',
 PosixPath('intervals_folder/0002-scattered.interval_list'),
 '-I',
 PosixPath('intervals_folder/0003-scattered.interval_list')]
```

Reply:
It's good as it is!

Reply:
Replaced it already because it is more readable.